python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) 2020-2023, NVIDIA CORPORATION.
+from typing import Any, Dict
+
+import numba
+import pandas as pd
+from numba import cuda, types
+from numba.core.extending import (
+    make_attribute_wrapper,
+    models,
+    register_model,
+    type_callable,
+    typeof_impl,
+)
+from numba.core.typing import signature as nb_signature
+from numba.core.typing.templates import AbstractTemplate, AttributeTemplate
+from numba.cuda.cudadecl import registry as cuda_registry
+from numba.np import numpy_support
+
+index_default_type = numpy_support.from_dtype(
+    pd.RangeIndex(0, 0).dtype
+)  # int64
+SUPPORTED_GROUPBY_NUMBA_TYPES = [types.int64, types.float64]
+SUPPORTED_GROUPBY_NUMPY_TYPES = [
+    numpy_support.as_dtype(dt) for dt in [types.int64, types.float64]
+]
+
+
+class Group(object):
+    """
+    A piece of python code whose purpose is to be replaced
+    during compilation. After being registered to GroupType,
+    serves as a handle for instantiating GroupType objects
+    in python code and accessing their attributes
+    """
+
+    def __init__(self, group_data, size, index, dtype, index_dtype):
+        self.group_data = group_data
+        self.size = size
+        self.index = index
+        self.dtype = dtype
+        self.index_dtype = index_dtype
+
+
+class GroupType(numba.types.Type):
+    """
+    Numba extension type carrying metadata associated with a single
+    GroupBy group. This metadata ultimately is passed to the CUDA
+    __device__ function which actually performs the work.
+    """
+
+    def __init__(self, group_scalar_type, index_type=index_default_type):
+        self.group_scalar_type = group_scalar_type
+        self.index_type = index_type
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    """
+    Tie Group and GroupType together such that when Numba
+    sees usage of Group in raw python code, it knows to
+    treat those usages as uses of GroupType
+    """
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )
+
+
+# The typing of the python "function" Group.__init__
+# as it appears in python code
+@type_callable(Group)
+def type_group(context):
+    def typer(group_data, size, index):
+        if (
+            isinstance(group_data, types.Array)
+            and isinstance(size, types.Integer)
+            and isinstance(index, types.Array)
+        ):
```

Comment:
Is the expected behavior of this function to return `None` if the condition isn't met?

Reply:
yes! This signals to the numba typing pass that an overload was not found.